File: simulate-browser/post_process.go
package simulate_browser

import (
	"log"
	"regexp"
)

// The Find/Replace values for post-processing the downloaded item.
type FindReplace struct {
	Find    *regexp.Regexp
	Replace string
}

// Apply the regular expressions in the list to the string.
func ApplyRegex(regexList []FindReplace, text string, verbose bool) string {
	logs := log.Default()
	for i := 0; i < len(regexList); i++ {
		if verbose {
			logs.Printf("Replacing '%s' with '%s'...\n", regexList[i].Find, regexList[i].Replace)
		}
		text = regexList[i].Find.ReplaceAllString(text, regexList[i].Replace)
	}
	return text
}

File: simulate-browser/post_process.cpp
#include "post_process.hpp"
#include <numeric>
#include <format>
#include <iostream>

std::string apply_regex(const std::vector<find_replace>& regex_list, const std::string& string, bool verbose) {
	return std::accumulate(regex_list.cbegin(), regex_list.cend(), string, [verbose](const std::string& accum, const find_replace& fr) {
		if (verbose) {
			std::cerr << std::format("Replacing '{}' with '{}'...\n", fr.find, fr.replace);
		}
		try {
			return std::regex_replace(accum, std::regex(fr.find), fr.replace);
		} catch (const std::exception& err) {
			throw std::runtime_error(std::format("Unable to apply regex: {}", err.what()));
		}
	});
}
